src/psyclone/tests/psyir/transformations/transformations_test.py

```diff
@@ -326,6 +327,86 @@ def test_omplooptrans_properties():
             in str(err.value))
 
 
+def test_parallellooptrans_validate_dependencies(fortran_reader):
+    ''' Test that the parallellooptrans validation checks for loop carried
+    dependencies. '''
+
+    # Use OMPLoopTrans as a concrete class of ParallelLoopTrans
+    omplooptrans = OMPLoopTrans()
+    # Example with a loop carried dependency in jk dimension
+    psyir = fortran_reader.psyir_from_source('''
+    subroutine my_subroutine()
+        integer :: ji, jj, jk, jpkm1, jpjm1, jpim1
+        real, dimension(10, 10, 10) :: zwt, zwd, zwi, zws
+        do jk = 2, jpkm1, 1
+          do jj = 2, jpjm1, 1
+            do ji = 2, jpim1, 1
+              zwt(ji,jj,jk) = zwd(ji,jj,jk) - zwi(ji,jj,jk) * &
+                              zws(ji,jj,jk - 1) / zwt(ji,jj,jk - 1)
+            enddo
+          enddo
+        enddo
+    end subroutine''')
+    loops = psyir.walk(Loop)
+
+    # Check that the loop can not be parallelised due to the loop-carried
+    # dependency.
+    with pytest.raises(TransformationError) as err:
+        omplooptrans.validate(loops[0])
+    assert ("Transformation Error: Dependency analysis failed with the "
+            "following messages:\nWarning: Variable 'zwt' is written and "
+            "is accessed using indices 'jk - 1' and 'jk' and can therefore "
+            "not be parallelised." in str(err.value))
+
+    # However, the inner loop can be parallelised because the dependency is
+    # just with 'jk' and it is not modified in the inner loops
+    omplooptrans.validate(loops[1])
+
+    # Check if there is missing symbol information it still validates
+    del loops[1].ancestor(Routine).symbol_table._symbols['zws']
+    omplooptrans.validate(loops[1])
+
+    # Reductions have dependencies but these are accepted because it
+    # can be manage with the appropriate clause
+    psyir = fortran_reader.psyir_from_source('''
+    subroutine my_subroutine()
+        integer :: ji, jj, jk, jpkm1, jpjm1, jpim1
+        real, dimension(10, 10, 10) :: zwt
+        real :: total
+        do jk = 2, jpkm1, 1
+          do jj = 2, jpjm1, 1
+            do ji = 2, jpim1, 1
+              total = total + zwt(ji,jj,jk)
+            enddo
+          enddo
+        enddo
+    end subroutine''')
+    loops = psyir.walk(Loop)
+    assert not DependencyTools().can_loop_be_parallelised(
+                    loops[0], only_nested_loops=False)
+    omplooptrans.validate(loops[0])
+
+    # Shared scalars are race conditions but these are accepted because it
+    # can be manage with the appropriate clause
+    psyir = fortran_reader.psyir_from_source('''
+    subroutine my_subroutine()
```

Comment:
You could template this code rather than repeat it - it's only the loop body that changes each time.

Reply:
Done